### DISCUSSION TITLE
Object Browser

### DISCUSSION THREAD

**Discussion by wqweto** [twinbasic/twinbasic#1212] [General]
As far I understand VBx **Object Browser** will be replaced by **References** node in the project tree view as mentioned [here](https://github.com/twinbasic/twinbasic/issues/163#issuecomment-849638577).

I cannot find ETA for this on the [roadmap issue](https://github.com/twinbasic/twinbasic/issues/335), is this feature abandoned?

Is there going to be some other impl for (referenced) typelibs viewer in the IDE?

**Comment by WaynePhillipsEA** [comment]
I think we'll definitely be wanting an Object Browser.

I've been considering, perhaps as an (easy to implement) initial alternative, perhaps we could have the references node list each reference, and then clicking on the individual reference displays a document containing the tB equivalent definitions of each type... like the IDL source, but with tB syntax.  That could be put together pretty quickly.

**Reply by wqweto** [reply]
Yes, that would be even better.

In C# projects I actually never use Object Browser but what I'm missing is F12 on foreign (typelib provided) data-types. Jumping into source dump would reduce cognitive load as it's still source code, not a different specialized UI as with Object Browser.

**Reply by WaynePhillipsEA** [reply]
I thought about this some more, and decided to implement it now rather than later, as it will help me considerably to not have to switch to OLEVIEW all the time, and it also helps to visually check we're picking everything up correctly out of the type libraries.

So we've got an first implementation of this built in to BETA 136.  Goto definition should work with the external type libraries now.  Once you're within the generated type library virtual document, Goto definition doesn't work inside there yet.   At the moment all references (included imported ones from other references) get picked up and you will see them in the `References` node of the Project explorer.
<img width="840" alt="tbTypeLibViewer" src="https://user-images.githubusercontent.com/25738109/193638531-799d5ad8-e07c-432e-9abf-1b8344209a0e.png">

**Reply by WaynePhillipsEA** [reply]
The actual formatting of the generated document is not final, but it gives you a good idea of how things will look once completed.

**Reply by DavidHollman** [reply]
@wqweto  if you're referring to Visual Studio (2019), for me I find that F12 on a type from a typelib does work - it shows a "decompiled" view of the TLB contents as though it were C#... I use that all the time.

I agree that presenting the TLB contents in this way is much easier to use/understand. Aside from cognitive load being less its also just easier & more consistent to access it since it uses the same mechanism (F12 = go to definition).

**Comment by fafalone** [comment]
If it's already in tB syntax this will be such a big time saver for converting the rest of oleexp... all I'd have to do is review types for pointers/handles.

**Comment by DavidHollman** [comment]
I was just playing with beta 136, here are some thoughts & notes about how the References node works relative to this "object browser" feature.

First off, I like this **much much more** than the original Object Browser. Having it in tB syntax is way more clear, as was noted above by @wqweto already.

Having the references show up exclusively as though they were twin files... maybe those nodes will need to be extended for other uses later? I could imagine there could be other properties or details that might fit into this spot in the UI but not necessarily into the decompiled sources. Maybe the decompiled TLB should eventually be a sub-node of a more abstract references node?

I like seeing the explicit references to stdole-based interfaces in the decompilation. This was always a "hidden" detail in VB6 but I think showing it is more powerful and demystifies things. 

Will any type of non-COM references show up under the references node? Like other tB projects? If so, what will be shown for them? (Or will those actually still use COM as the only interface?)

What about DLLImport style "references" (dependencies). Can they show up here too? It could just be a list of exports. (In which case its OLEView + Dependency Walker together which would be really powerful).

Minor issues:

- I think showing the DLL name as the node name instead of the .twin pseudo filename would be more clear. e.g., show `MSXML2.dll` instead of `MSXML2.twin` in the tree.

-  <strike>_Go to type definition_ does not seem to work from within one of the decompiled files. Even when the definition is in the same file itself. For example, open stdole.twin, find `IDispatch` which extends from `stdole.IUnknown`. Right click `stdole.IUnknown` in that spot and select "go to type definition" and it just gives the error "no type definition found".</strike>  (Noted above in https://github.com/twinbasic/twinbasic/discussions/1212#discussioncomment-3789300 which I overlooked)

- What does it mean when something extends from `Nothing`?  For example: `Interface IUnknown Extends Nothing`. 

- I opened a VBP for a DLL project. It references both VB6 and C# DLLs. _At first the C# references all showed up blank_ in the right hand side panel, though the VB6 ones worked immediately. Some time later I noticed the C# ones were working however. A delay in loading them?
 
- The _ImportedTypeLibraries_ node a the top of the project tree is empty. Is it needed anymore?

- There seems to be extra indentation after comments and maybe a missing CRLF. Example:

```
Interface IXMLDOMNode Extends stdole.IDispatch
    /* voffset &H001C */ [ DispId(2) ]
        [ Description ("name of the node") ]
        Property Get nodeName() As String

```

could be:

```
Interface IXMLDOMNode Extends stdole.IDispatch
    /* voffset &H001C */ 
    [ DispId(2) ]
    [ Description ("name of the node") ]
    Property Get nodeName() As String

```

- Maybe allowing saving the decompiled sources directly from the IDE would be useful. Right now the save option is disabled.

- The decompiled VBA.twin file has some strange looking lines near the top:
```
Alias VbVarType As 

Alias VbMsgBoxStyle As 

Alias VbMsgBoxResult As 

Alias VbFileAttribute As 

Alias VbStrConv As 

Alias VbDayOfWeek As 

Alias VbFirstWeekOfYear As 

Alias VbIMEStatus As 

Alias VbAppWinStyle As 

Alias VbCompareMethod As 

Alias VbCalendar As 

Alias VbDateTimeFormat As 

Alias VbTriState As 

Alias VbCallType As 

DataType 

DataType 

DataType 

DataType 

DataType 

DataType 

DataType 

DataType 

DataType 

DataType 

DataType 

DataType 

DataType 

Alias _Collection As 

Alias Collection As 

```

I hadn't noticed that in any other references.

- I think adding the source .DLL filename & path to the header of the decompiled sources would be very helpful, especially when tracking down DLL hell sorts of issues and you have multiple copies of the same file.

---

This was just a brain dump... it got quite long. If it would help if I split any of these into separate issues please just let me know.

**Reply by DavidHollman** [reply]
@WaynePhillipsEA  you dealt with most of the suggestions I made above in 137, thanks! :)

**Comment by rexxitall** [comment]
For all those viewers also the viewer of the proect references a text filter would be nice. I hate it if i add new references and have to read and watch hundreds of entrys. Some kind of text entry field orcombobox where you can add part of the name, "mic" for all microsoft entrys as example. If they are tree views the should be shown collapsed as default.

**Comment by fafalone** [comment]
IUnknown is the root interface from which all others inherit; Extends Nothing was special syntax for that special case where there is nothing to inherit from because it is the root.

VB placed a lot of restrictions on using it directly, so a few of us made custom declared versions for when we wanted to access it's methods directly, like to increment the reference counter with .AddRef so it won't be freed in some rare circumstances where VB otherwise would and that isn't desirable.

You aren't likely to to ever see it besides for IUnknown, things won't work right if you use it for anything else unless you replicated it's methods manually anyway.

**Comment by WaynePhillipsEA** [comment]
In BETA 137... lots of improvements have now gone into the type library viewer, including F12 / Goto definition support within the type library documents themselves.

**Comment by fafalone** [comment]
After a couple days I have to say seeing all the 'tB style' syntax of things we can consume but not implement in language yet (coclass, [ PreserveSig ], Alias....)... it's a huge tease :D

**Comment by Kr00l** [comment]
I don't know if the following duplicate which I noticed is intended or not.. so I let you decide.

This tlb interface:
```
    [
      odl,
      uuid(38584260-0CFB-45E7-8FBB-5D20B311F5B8),
      version(1.0),
      dual,
      nonextensible,
      oleautomation
    ]
    interface IOleInPlaceActiveObjectVB : IDispatch {
        [id(0x00000001)]
        HRESULT TranslateAccelerator(
                        [in, out] VARIANT_BOOL* Handled, 
                        [in, out] long* RetVal, 
                        [in] long hWnd, 
                        [in] long wMsg, 
                        [in] long wParam, 
                        [in] long lParam, 
                        [in] long Shift);
    };
```

is twice in tB reference page:

1.
```
    [ InterfaceId ("38584260-0CFB-45E7-8FBB-5D20B311F5B8") ]
    [ DualInterface, DispInterface]
    Interface IOleInPlaceActiveObjectVB Extends OLEGuids.IDispatch
        [ DispId(1) ]
        /* voffset &H001C*/ Sub TranslateAccelerator(ByRef Handled As Boolean, ByRef RetVal As Long, ByVal hWnd As Long, ByVal wMsg As Long, ByVal wParam As Long, ByVal lParam As Long, ByVal Shift As Long)
    End Interface
```

2.
```
    [ InterfaceId ("38584260-0CFB-45E7-8FBB-5D20B311F5B8") ]
    [ DualInterface]
    Interface IOleInPlaceActiveObjectVB Extends OLEGuids.IDispatch
        [ DispId(1) ]
        /* voffset &H001C*/ Sub TranslateAccelerator(ByRef Handled As Boolean, ByRef RetVal As Long, ByVal hWnd As Long, ByVal wMsg As Long, ByVal wParam As Long, ByVal lParam As Long, ByVal Shift As Long)
    End Interface
```

Difference is only that once with `[ DualInterface, DispInterface]` and again with only `[ DualInterface]`

**Reply by WaynePhillipsEA** [reply]
Internally we generate both sides of the dual interface (both the dispinterface, and the vtable interface), so that's where this comes from.  But in terms of what you get to see in the type library viewer, I'd call it a bug.
